p4-16/spec/P4-16-spec.mdk

```diff
@@ -4180,11 +4180,38 @@ h = { 10, 12 };  // This also makes the header h valid
 h = { y = 12, x = 10 };  // Same effect as above
 ~ End P4Example
 
-Two headers can be compared for equality (==) or inequality (!=) only
+Two headers can be compared for equality (`==`) or inequality (`!=`) only
 if they have the same type.  Two headers are equal if and only if they
 are both invalid, or they are both valid and all their corresponding
 fields are equal.
 
+The expression `{#}` represents an invalid header of some type, but it
+can be any header or header union type.  A P4 compiler may require an
+explicit cast on this expression in cases where it cannot determine
+the particular header or header union type from the context.
```

Comment:
If the type of '{#}' cannot be inferred from the context the program will be rejected.

Reply:
@mbudiu-vmw  Thanks for taking a look.  Just so I am clear, you are recommending replacing my sentence "A P4 compiler may require an explicit cast on this expression in cases where it cannot determine the particular header or header union type from the context." with the text of your comment?  I am fine doing that, if so.

Reply:
The main question is whether that is a cast or an entire new expression.
The implementation treats it as a cast and the converts it into a new expression.